### ISSUE TITLE
LFX Workspace: OpenCV SDKs for Wasm

### ISSUE THREAD

**Issue by omkar-mohanty** [WasmEdge/WasmEdge#1836]
# Motivation
WasmEdge is a leading WebAssembly runtime for AI inference. It supports AI frameworks such as Tensorflow, OpenVINO and PyTorch. A compelling use case is computer vision applications on the edge.

Computer vision applications need to pre-process images and videos into tensor formats before applying the AI model. They then often need to overlay the tensor results onto the original image. In our existing demos, we use the Rust [image crate](https://crates.io/crates/image) to process images. However, the crate only has limited features and is inadequate for many computer vision applications.

In the Python-based computer vision applications, the image pre-processing is often done with the Python wrapper for OpenCV library. The OpenCV library itself is written in C and can be compiled into WebAssembly.

# Schedule 
- [x] Compile [Rust wrapper](https://github.com/twistedfall/opencv-rust) crate into `wasm32` by fixing  errors, 3 weeks.
- [x] Write Rust examples on using opencv in WasmEdge, 1 week.
- [ ] Create and test JavaScript wrapper for opencv in WasmEdge QuickJS, 3 week.
- [ ] Write JavaScript examples on using opencv in WasmEdge, 2 week.
- [ ] Benchmark opencv running on WasmEdge against Python version, 1 week.

# Appendix
- #1747

**Comment by juntao** [comment]
Thank you. 

I think you will need to create the JavaScript wrapper API for QuickJS first before you can write the JavaScript examples.

I would also like to know a more detailed task breakdowns for the first item (compiling the rust wrapper). What specifically do you need to fix to compile it to wasm32-wasi? Can we contribute the fixes upstream? Thanks!

**Comment by omkar-mohanty** [comment]
> Thank you.
> 
> I think you will need to create the JavaScript wrapper API for QuickJS first before you can write the JavaScript examples.
> 
> I would also like to know a more detailed task breakdowns for the first item (compiling the rust wrapper). What specifically do you need to fix to compile it to wasm32-wasi? Can we contribute the fixes upstream? Thanks!

Hi @juntao ,
I fixed the first one, I slightly overlooked the order of tasks.

About the first item I found out that the rust wrapper crate compiles when I set  `--target=wasm32-unknown-emscripten` but `clang` errors occour when I set `--target=wasm32-wasi`. 
```
error occurred: Command "clang" "-O0" "-ffunction-sections" "-fdata-sections" "-fPIC" "-g" "-fno-omit-frame-pointer" "--target=wasm32-wasi" "-I" "/home/omkar/Source/opencv-rust/src_cpp" "-I" "/home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-6b60cea97f53d006/out" "-I" "." "-I" "/usr/local/include/opencv4" "-Wall" "-Wextra" "-std=c++11" "-o" "/home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-6b60cea97f53d006/out/calib3d.o" "-c" "/home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-6b60cea97f53d006/out/calib3d.cpp" with args "clang" did not execute successfully (status code exit status: 1).
````
I am still researching what exactly went wrong.

**Comment by omkar-mohanty** [comment]
After tinkering around the rust wrapper crate I successfully compiled to `wasm32-unknown-emscripten` but when compiling it to `wasm32-wasi` the clang compiler fails to do so and is not very clear where it went wrong.

**Comment by juntao** [comment]
Have you followed the examples of WasmEdge QuickJS? It was the same approach.

**Comment by omkar-mohanty** [comment]
I tried compiling it the way it was in QuickJS but it didn't work as I mentioned above. After looking at the docs for [wasm32-wasi target](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_target/spec/wasm32_wasi/index.html), I figured that I need to compile with `-C target-feature=-crt-static`  and also provide the `-C linker` to clang binary in [wasi-sdk](https://github.com/WebAssembly/wasi-sdk) but I still got some trouble compiling `memchar` and `libc`.  Mostly due to wrong flags passed to `wasm-ld`.

**Comment by juntao** [comment]
What are the error messages? Can you write down how to re-produce your work so far? Thanks.

**Comment by juntao** [comment]
Can you provide the complete steps for re-producing the error? Start from how you get OpenCV source code and list your build environments (OS, CPU, Rust version etc). Thanks.

**Comment by omkar-mohanty** [comment]
## Specifications 
OS : Kubuntu 22.04.1 LTS x86_64
CPU : Intel i7-1065G7 (8) @ 3.900GHz 
Rust : 1.66.0-nightly
RAM : 16 GB
GPU : Nvdia MX 330 with Intel iris integrated
## Steps to reproduce the error 
### 1. OpenCV
```bash
sudo apt-get install libopencv-dev
```
### 2. For verifying OpenCV
```bash 
pkg-config --modversion opencv4
```
### 3. Wasi-sdk Installation
Needed for sysroot 
```bash 
export WASI_VERSION_FULL=${WASI_VERSION}.0
wget https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-${WASI_VERSION}/wasi-sdk-${WASI_VERSION_FULL}-linux.tar.gz
tar xvf wasi-sdk-${WASI_VERSION_FULL}-linux.tar.gz
mv wasi-sdk-${WASI_VERSION_FULL} wasi-sdk
sudo mv wasi-sdk /opt/
```
### 4. Rust wrapper crate
```bash 
git clone https://github.com/twistedfall/opencv-rust.git
cd opencv-rust 
export CC_wasm32_wasi=/opt/wasi-sdk/bin/clang
export CARGO_TARGET_WASM32_WASI_LINKER=/opt/wasi-sdk/bin/clang
export RUSTFLAGS='-C target-feature=-crt-static -C linker=/opt/wasi-sdk/bin/clang -C link-args=-Wl,-v,--no-entry,--export-all'
cargo build --target=wasm32-wasi
```

**Comment by juntao** [comment]
@omkar-mohanty thank you. 

As you can see from the commit below, the QuickJS project required a few changes to add `wasi` as a conditional platform for the compiler. I think that you need to do the same for this project. Thanks. 

https://github.com/second-state/quickjs-wasi/commit/baecae488a972604f232cb68111ff0df80734ac0

**Comment by omkar-mohanty** [comment]
I looked at the commit and did some further research and found that our compilation is related to this issue https://github.com/rust-lang/rust/issues/76679. I did some tinkering around the `wasm-ld` and did away from the aforementioned errors but now there are some linking errors from `wasm-ld`

```bash 
wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.0.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.1.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.10.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.11.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.12.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.13.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.14.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.15.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.2.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.3.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.4.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.5.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.6.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.7.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.8.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.build_script_build.5f6ca34a-cgu.9.rcgu.o
            wasm-ld: error: unknown file type: /home/omkar/Source/opencv-rust/target/wasm32-wasi/debug/build/opencv-00430ce37c3eeed3/out/release/build/libc-89a29c9379dcf2b2/build_script_build-89a29c9379dcf2b2.4r45rigk85cqimyu.rcgu.o
            wasm-ld: error: unable to find library -lgcc_s
            wasm-ld: error: unable to find library -lutil
```
The build is same as before.

**Comment by omkar-mohanty** [comment]
@juntao I've successfully compiled OpenCV to wasm32-wasi. Now to use OpenCV in wasmedge i need to generate wrappers for rust one way is to use bindgen tool but it doesn't fully support c++, second option is to try to use the rust wrapper crate's binding generator which wraps all c++ into C API.

**Comment by juntao** [comment]
Great. Can you provide the detailed instructions on how to compile OpenCV to WASI so that we can re-produce on our end?

For the Rust wrapper for C++, does this help? https://docs.rs/cxx/latest/cxx/

**Comment by omkar-mohanty** [comment]
@juntao I made a [shell script](https://github.com/omkar-mohanty/opencv-wasm/tree/main) which should compile OpenCV to WASI.

**Comment by juntao** [comment]
Ok. Thanks!

So you do not need to make any changes to the opencv source code for it to compile for wasi?

**Comment by omkar-mohanty** [comment]
Not really the errors I was getting from above messages were from compiling the rust wrapper crate. I wasted a lot of time trying to compile the wrapper crate to WASI only to discover that it does not support Wasm target https://github.com/twistedfall/opencv-rust/issues/124

**Comment by omkar-mohanty** [comment]
@juntao I created the rust project using CXX crate and following the quickJS project. Here is the [link](https://github.com/omkar-mohanty/opencv-wasm), it contains some functions from `imgproc` and `imgcodecs` for now. However while building the crate the `wasm-ld` linker fails to find `libstdc++` for some reason I tried looking around to how to fix it but couldn't find any. Thanks

**Comment by juntao** [comment]
Thanks. @L-jasmine can you look into this? Thanks. 

> @juntao I created the rust project using CXX crate and following the quickJS project. Here is the [link](https://github.com/omkar-mohanty/opencv-wasm), it contains some functions from `imgproc` and `imgcodecs` for now. However while building the crate the `wasm-ld` linker fails to find `libstdc++` for some reason I tried looking around to how to fix it but couldn't find any. Thanks

**Comment by omkar-mohanty** [comment]
@juntao I fixed all the issues.Rust examples of Image reading  and writing are working (for jp2 format at the moment). All the code is in the [repo](https://github.com/omkar-mohanty/opencv-wasm)

**Comment by juntao** [comment]
Hi @omkar-mohanty 

Can you summarize the current status of this project?

1 Can we compile the entire OpenCV library into WASI?

2 Which OpenCV functions are currently available in the Rust wrapper for WebAssembly?

3 How much effort is needed to expose the rest of OpenCV functions in Rust? Are their any technical blockers? 

Thanks

**Comment by omkar-mohanty** [comment]
1. Not the entire library , for now I have compiled `imgcodecs` ,`imgproc` and `core` to WASI. 
2. From `imgcodecs` module:-
- `imread`
- `imwrite`
   From `imgproc` module:-
- `bilateralFilter`
- `boxFilter`
- `blur`
3. The main technical blocker for now is lack of exception support and thread support as we would have to modify the source code every time we encounter exceptions and threads.

**Comment by juntao** [comment]
Hmm. OpenCV can be compiled to Wasm for the browser. Do they also need to make code changes for it to compile? Have you tried this to disable threads?

https://docs.opencv.org/3.4/db/de0/group__core__utils.html#gae78625c3c2aa9e0b83ed31b73c6549c0

**Comment by ChillSheep** [comment]
Is there any work on this still being done?

**Comment by hydai** [comment]
Hi @ChillSheep 
We have a new PR for this. https://github.com/WasmEdge/WasmEdge/pull/2648
Also, it would be a new issue to cover in the next LFX mentorship project.

**Comment by dannypsnl** [comment]
Close since we already have #2648 merged, as above.
